theseus/src/launcher/download.rs

```diff
@@ -1,362 +1,222 @@
+//! Downloader for Minecraft data
+
 use crate::{
-    data::{DataError, Settings},
-    launcher::LauncherError,
-};
-use daedalus::get_path_from_artifact;
-use daedalus::minecraft::{
-    fetch_assets_index, fetch_version_info, Asset, AssetsIndex, DownloadType,
-    Library, Os, Version, VersionInfo,
-};
-use daedalus::modded::{
-    fetch_partial_version, merge_partial_version, LoaderVersion,
+    state::DirectoryInfo as Dirs,
+    util::{fetch::*, platform::OsExt},
 };
-use futures::future;
-use std::path::Path;
-use std::time::Duration;
-use tokio::{
-    fs::File,
-    io::AsyncWriteExt,
-    sync::{OnceCell, Semaphore},
+use daedalus::{
+    self as d,
+    minecraft::{
+        Asset, AssetsIndex, Library, Os, Version as GameVersion,
+        VersionInfo as GameVersionInfo,
+    },
+    modded::LoaderVersion,
 };
+use futures::prelude::*;
+use tokio::fs;
 
-static DOWNLOADS_SEMAPHORE: OnceCell<Semaphore> = OnceCell::const_new();
```

Comment:
Without the semaphore are we just doing as many concurrent HTTP requests? Could we bring back the functionality of having the number of concurrent requests being configurable (useful for slow connections)?

Reply:
This is now stored in the global state, and we plan to merge them by passing in the permit.